File: GitToolBox/src/main/java/zielu/gittoolbox/revision/RevisionService.java
package zielu.gittoolbox.revision;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import zielu.gittoolbox.util.AppUtil;

public interface RevisionService {
  static RevisionService getInstance(@NotNull Project project) {
    return AppUtil.getServiceInstance(project, RevisionService.class);
  }

  @NotNull
  RevisionInfo getForLine(@NotNull RevisionDataProvider provider, int lineNumber);

  @Nullable
  String getCommitMessage(@NotNull VirtualFile file, @NotNull RevisionInfo revisionInfo);
}

File: src/main/kotlin/zielu/gittoolbox/revision/RevisionService.kt
package zielu.gittoolbox.revision

import com.google.common.cache.CacheBuilder
import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.vcs.history.VcsRevisionNumber
import com.intellij.openapi.vfs.VirtualFile
import zielu.gittoolbox.util.AppUtil
import java.time.Duration
import java.util.concurrent.ExecutionException

internal class RevisionService(
  private val project: Project
) : Disposable {
  private val commitMessageCache = CacheBuilder.newBuilder()
    .maximumSize(50)
    .expireAfterAccess(Duration.ofMinutes(30))
    .recordStats()
    .build<VcsRevisionNumber, String>()
  private val facade = RevisionServiceFacade(project)
  init {
    facade.exposeCommitMessageCacheMetrics(commitMessageCache)
  }

  override fun dispose() {
    commitMessageCache.invalidateAll()
  }

  fun getForLine(provider: RevisionDataProvider, lineNumber: Int): RevisionInfo {
    return RevisionInfoFactory.getInstance(project).forLine(provider, lineNumber)
  }

  fun getCommitMessage(file: VirtualFile, revisionInfo: RevisionInfo): String? {
    if (revisionInfo.isEmpty()) {
      return null
    }
    val revisionNumber = revisionInfo.getRevisionNumber()
    return try {
      commitMessageCache[revisionNumber, { facade.loadCommitMessage(file, revisionNumber) }]
    } catch (e: ExecutionException) {
      log.warn("Failed to load commit message $revisionNumber", e)
      null
    }
  }

  companion object {
    private val log = Logger.getInstance(RevisionService::class.java)

    @JvmStatic
    fun getInstance(project: Project): RevisionService {
      return AppUtil.getServiceInstance(project, RevisionService::class.java)
    }
  }
}
